### DISCUSSION TITLE
Is there a method which would "columnize" an IEnumerable?

### DISCUSSION THREAD

**Discussion by andlehma** [morelinq/MoreLINQ#823] [Q&A]
For example, if I were writing a Tic Tac Toe implementation which takes an input IEnumerable where the indices are mapped like:
0 | 1 | 2
3 | 4 | 5
6 | 7 | 8

Rows are easy to check, but I don't know of a single-method way to get the columns. I have settled on the following:
```cs
var columns = gameState.Select((val, idx) => new { index = idx, value = val})
  .GroupBy(x => x.index % 3)
  .Select(x => x.Select(y => y.value));
```
I wonder if there is a cleaner way to do this. LINQ's GroupBy does not have access to the index of each element, so I can't do `index % 3` or something similar. I'm not sure how frequent of a use case this is, but it might be worth considering adding a `Columnize` method to MoreLINQ, or maybe just a GroupBy method which exposes the index of each elemenet.

**Comment by viceroypenguin** [comment]
[`.Batch()`](https://github.com/morelinq/MoreLINQ/blob/cae4bcce7b12e7e4400dd21c803c1850dbdbaca8/MoreLinq/Batch.cs#L51) would do the trick.

In this case, it would be `var columns = gameState.Batch(3);`

**Comment by andlehma** [comment]
Actually, `Batch` gives me the rows. I'm looking for the columns, i.e. indices 0, 3, 6 would be one column, indices 1, 4, 7 another column.

**Comment by viceroypenguin** [comment]
Oh! Then yes, I believe your code would be the simplest way that does a one-time pass through the enumerable. You could use `.TakeEvery()` and use a separate pass for each column, but I would probably recommend what you have instead.

**Comment by atifaziz** [comment]
I believe you're looking for [`Transpose`](http://morelinq.github.io/3.3/ref/api/html/M_MoreLinq_MoreEnumerable_Transpose__1.htm):

```c#
const string gameStateString = @"
    0 | 1 | 2
    3 | 4 | 5
    6 | 7 | 8 ";

var gameState =
    from m in Regex.Matches(gameStateString, @"[0-9]+")
    select int.Parse(m.Value);

var rows = gameState.Batch(3);
var columns = rows.Transpose();
```
